File: evaluate/SMA.py
from numpy import double
import logging

from Config import SMA_MIN, SMA_STEP, BUY_THRESHOLD, SELL_THRESHOLD, TRADING_FEE, DOWNTIME_DAYS, TRADE_MODE


class SMA:
    """Represents a single Simple Moving Average trading strategy with buy/sell logic."""

    days = 0
    bought = False
    buyPrice = double(0.0)
    totalProfit = double(0.0)
    downtimeDays = 0
    smaMark = double(0.0)

    def __init__(self, days):
        self.days = days
        self.bought = False
        self.buyPrice = double(0.0)
        self.totalProfit = double(0.0)
        self.downtimeDays = 0
        self.smaMark = double(0.0)

    def smaUpdate(self, updater):
        """Active every trade day at end - update SMA mark from file."""
        # Note: SMA data file is updated once per day by Evaluator. Just read it here.

        # read SMA from file - calculate line index based on SMA_MIN and SMA_STEP
        try:
            readFile = open("SMA.txt", "r")
            # Calculate which line this SMA is on (0-indexed)
            lineIndex = (self.days - SMA_MIN) // SMA_STEP
            
            # Read to the correct line
            for _ in range(lineIndex):
                readFile.readline()

            smaValue = readFile.readline().strip()
            readFile.close()
            
            if smaValue == "" or smaValue == "NaN":
                logging.warning(f"Invalid SMA value for SMA {self.days}: '{smaValue}'")
                # Keep previous value
                return
            
            self.smaMark = double(smaValue)

        except Exception as e:
            logging.error(f"Error trying to read SMA data for SMA {self.days}: {e}")
            raise e

    def report(self, logger):
        """Send profits so far to logger for printing."""
        logger.giveEvalReport(self.days, self.totalProfit)

    def force_liquidate(self, price, logger):
        """Force-liquidate any open position at the provided price.

        Returns True if a position was closed, otherwise False.
        Writes a unique message to the evaluation log for easy grep/analysis.
        """
        if self.bought:
            profit = ((price - self.buyPrice) - TRADING_FEE)
            self.totalProfit = self.totalProfit + profit
            tempProfit = double(profit)
            # Unique marker: FORCE-LIQUIDATED
            logger.appendToEvalLog(
                f"SMA bot {self.days} Force-Liquidated at {price} for net of {tempProfit}. SMA: {self.smaMark}."
            )
            # Reset state after liquidation
            self.bought = False
            self.buyPrice = double(0.0)
            return True
        return False

    def smaDowntimeUpdate(self):
        """Starts every day - decrement downtime counter."""
        if self.downtimeDays != 0:
            self.downtimeDays -= 1

    def smaAction(self, price, logger):
        """Active every time slice during trade time - buy/sell logic."""
        # Determine buy/sell signals based on TRADE_MODE
        if TRADE_MODE == 'mean_reversion':
            # mean reversion: buy when price is sufficiently below SMA, sell when above
            buy_condition = (not self.bought) and (self.downtimeDays == 0) and (price + BUY_THRESHOLD < self.smaMark)
            sell_condition = self.bought and (price > self.smaMark + SELL_THRESHOLD)
        else:
            # default/momentum: buy when price sufficiently above SMA, sell when below
            buy_condition = (not self.bought) and (self.downtimeDays == 0) and (self.smaMark + BUY_THRESHOLD < price)
            sell_condition = self.bought and (self.smaMark > price + SELL_THRESHOLD)

        if buy_condition:
            self.bought = True
            self.buyPrice = price
            logger.appendToEvalLog(
                f"SMA bot {self.days} bought at {price}. SMA: {self.smaMark}."
            )
        if sell_condition:
            # perform sell
            self.downtimeDays = self.downtimeDays + DOWNTIME_DAYS
            self.bought = False
            self.totalProfit = self.totalProfit + ((price - self.buyPrice) - TRADING_FEE)
            tempProfit = double(((price - self.buyPrice) - TRADING_FEE))
            logger.appendToEvalLog(
                f"SMA bot {self.days} sold at {price} for a profit of {tempProfit}. SMA: {self.smaMark}."
            )

File: run/SMA.java
package run;

import com.sun.tools.javac.Main;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.util.Objects;

import static log.LogManager.*;

/**
 * Java implementation of SMA trading strategy.
 * TODO: Note from project - should be converted to Python.
 */
public class SMA {
    private final int days;
    private boolean bought = false;
    private double buyPrice;
    private double totalProfit = 0;
    private int downtimeDays = 0;
    private double SMAmark;

    public SMA(int days) {
        this.days = days;
    }

    /**
     * Active every trade day at end - updates SMA mark from Python script output.
     */
    public void SMAupdate() {
        // Get SMA from script
        try {
            // Run the python script
            ClassLoader classLoader = com.sun.tools.javac.Main.class.getClassLoader();
            File path = new File(Objects.requireNonNull(classLoader.getResource("python")).toURI());

            ProcessBuilder process = new ProcessBuilder();
            process.directory(new File(path.getAbsolutePath()));
            process.command("python3", "SMAupdater.py");
            process.start();

        } catch (Exception e) {
            System.out.println("Error grabbing SMA data for given ticker.");
            throw new RuntimeException(e);
        }

        // Read SMA from file
        try {
            ClassLoader classLoader = Main.class.getClassLoader();
            File path = new File(Objects.requireNonNull(classLoader.getResource("SMA.txt")).toURI());
            FileReader file = new FileReader(path);
            BufferedReader read = new BufferedReader(file);

            for (int i = 1; i < days; i++) {  // skip other SMAs
                read.readLine();
            }
            SMAmark = Double.parseDouble(read.readLine());  // get SMA
            read.close();
            file.close();
            // System.out.println(SMAmark);
        } catch (Exception e) {
            System.out.println("Error trying to read SMA data.");
            throw new RuntimeException(e);
        }
    }

    public void Report() throws IOException {
        GiveRunnerReport(days, totalProfit);  // sends profits so far and log prints it
    }

    /**
     * Starts every day - decrements downtime counter.
     */
    public void SMAdowntimeUpdate() {
        if (downtimeDays != 0) {
            downtimeDays -= 1;
        }
    }

    /**
     * Active every 30 min only on trade time - buy/sell logic.
     */
    public void SMAaction(double price) throws IOException {
        if (!bought) {  // buying
            if (downtimeDays == 0) {  // not in T-1 (2 days)
                if (SMAmark + 0.05 < price) {  // above SMA (by more than 0.05) -> buy
                    bought = true;
                    buyPrice = price;
                    AppendToRunLog("SMA bot " + days + " bought at " + price + ". SMA: " + SMAmark + ".");
                    System.out.println("SMA bot " + days + " bought at " + price + ". SMA: " + SMAmark + ".");
                }
            }
        } else {  // selling
            if (SMAmark > price + 0.05) {  // below SMA (by more than .05) -> sell
                downtimeDays = downtimeDays + 2;
                bought = false;
                totalProfit = totalProfit + ((price - buyPrice) - 0.05);  // .05 is fee
                double tempProfit = ((price - buyPrice) - 0.05);
                AppendToRunLog("SMA bot " + days + " sold at " + price + " for a profit of " + tempProfit + ". SMA: " + SMAmark + ".");
                System.out.println("SMA bot " + days + " sold at " + price + " for a profit of " + tempProfit + ". SMA: " + SMAmark + ".");
            }
        }
    }

    public int getDays() {
        return days;
    }

    public void setTotalProfit(double profit) {
        totalProfit = profit;
    }
}